policy/scan/pdque/queue.go

```diff
@@ -0,0 +1,414 @@
+//			return files[i].Name() < files[j].Name()
+//		})
+//
+//		for _, file := range files {
+//			if filepath.Ext(file.Name()) == jobFileExt {
+//				jobPath := filepath.Join(q.path, file.Name())
+//
+//				data, err := os.ReadFile(jobPath)
+//				if err != nil {
+//					return nil, err
+//				}
+//
+//				// Decode the bytes into an object
+//				obj := q.builder()
+//				if err := gob.NewDecoder(bytes.NewReader(data)).Decode(obj); err != nil {
+//					return nil, ErrUnableToDecode{
+//						Path: q.path,
+//						Err:  err,
+//					}
+//				}
+//
+//				// Remove the job file
+//				if err := os.Remove(jobPath); err != nil {
+//					return nil, err
+//				}
+//
+//				return obj, nil
+//			}
+//		}
+//
+//		// No jobs in queue, wait for a new job to be enqueued
+//		q.cond.Wait()
+//	}
+//}
+
+func (q *Queue) DequeueBlock() (interface{}, error) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	for {
+		if q.closed {
+			return nil, ErrQueueClosed
+		}
+
+		obj, err := q.dequeueJob()
+		if err != nil {
+			if errors.Is(err, ErrNoJobs) {
+				// No jobs in queue, wait for a new job to be enqueued
+				q.cond.Wait()
+			} else {
+				return nil, err
+			}
+		} else {
+			return obj, nil
+		}
+	}
+}
+
+// dequeueJob tries to dequeue a job from the queue without waiting.
+// It returns an ErrNoJobs error if there are no jobs to dequeue.
+func (q *Queue) dequeueJob() (interface{}, error) {
+	files, err := os.ReadDir(q.path)
+	if err != nil {
+		return nil, err
+	}
+
+	// Sort job files by name (which is the timestamp)
+	sort.Slice(files, func(i, j int) bool {
+		return files[i].Name() < files[j].Name()
+	})
+
+	for _, file := range files {
+		if filepath.Ext(file.Name()) == jobFileExt {
+			jobPath := filepath.Join(q.path, file.Name())
+
+			data, err := os.ReadFile(jobPath)
+			if err != nil {
+				return nil, err
+			}
+
+			// Decode the bytes into an object
+			obj := q.builder()
+			if err := gob.NewDecoder(bytes.NewReader(data)).Decode(obj); err != nil {
+				return nil, ErrUnableToDecode{
+					Path: q.path,
+					Err:  err,
+				}
+			}
+
+			// Remove the job file
+			if err := os.Remove(jobPath); err != nil {
+				return nil, err
+			}
+
+			return obj, nil
+		}
+	}
+
+	return nil, ErrNoJobs // Custom error to indicate no jobs are available
+}
+
+// ErrNoJobs is a custom error to indicate that no jobs are available in the queue.
+var ErrNoJobs = errors.New("no jobs in queue")
+
+func (q *Queue) nextAvailableFilename() (string, error) {
+	timestamp := time.Now().UnixNano()
+	filename := strconv.FormatInt(timestamp, 10)
+	for {
+		_, err := os.Stat(filepath.Join(q.path, filename+jobFileExt))
+		if os.IsNotExist(err) {
+			break
+		} else if err != nil {
+			return "", err
+		} else {
+			timestamp++
+			filename = strconv.FormatInt(timestamp, 10)
+		}
+	}
+	return filename, nil
+}
+
+// check the currentSize of the queue
+// We do a lot of disk operations here, could find a more performant approach
+func (q *Queue) currentSize() (int, error) {
```

Comment:
this is just for a follow-up: 
Do we expect multiple go-processes to run jobs on the same queue-folder? If so, then yes, we'd have to check the folder contents on every call. If not, we could check the contents once at startup and then just keep track of the internal # of jobs. (my general recommendation is to avoid running parallel processes, but I'd love to learn more!)

Reply:
Also: there is an argument for blocking the queue while counting on disk (because jobs may be added in the meantime), but in the multi-process use-case it is almost impossible to safeguard since any other process will likely not block and wait

Reply:
I am not sure if we do multiple processes in production but the queue should be able to handle it if so, there is a test for it too. The `currentSize` function does not have a lock but it's only called by `Enqueue` which does have a lock - so I think we should be safe here. Question remains if the performance of this will eventually become a bottleneck or not.

Reply:
the current library makes sure you cannot start multiple processes with the same disk queue (folder)

Reply:
Awesome, thank you!!